checks-superstaq/checks_superstaq/license_header_format_.py

```diff
@@ -0,0 +1,546 @@
+                    license_header_lst[-1].end_line_num = line_num + 1
+                    license_header = ""
+                break
+
+            if all(exception not in line for exception in exceptions):
+                if not license_header:
+                    license_header_lst.append(LicenseHeader(line_num + 1))
+
+                if line == "\n":
+                    # set the line number for the last line of the license_header
+                    license_header_lst[-1].license_header = license_header.strip()
+                    license_header_lst[-1].end_line_num = line_num + 1
+                    license_header = ""
+                else:
+                    license_header += line
+    license_header_lst = [
+        header for header in license_header_lst if "Copyright" in header.license_header
+    ]
+
+    return license_header_lst
+
+
+def _validate_license_header(
+    license_header_lst: list[LicenseHeader],
+    expected_license_header: str,
+    licensee: str,
+    editable: bool,
+) -> bool:
+    """Returns whether there is a valid license header in a file and for each license
+    header in a file, it assigns each theiir type.
+        - VALID: if the header contains a Copyright line.
+        - OUTDATED: if the header has the correct licensee but different type.
+        - SIMILAR_LICENSE: if the header is a valid license but not from the licensee.
+        - OTHER: if the header is any other one.
+    Args:
+        license_header_lst: List of the license_header objects for each header in a file.
+        licensee: The owner of the expected license.
+        editable: Whether similar licenses can be edited to include the license owner instead
+            adding the entire license to the file.
+
+    Returns: Whether there is a valid license header in a file or not.
+    """
+    copyright_line = ""
+    body = ""
+    copyright_pattern = re.compile(r"Copyright .*")
+    header_as_lst = expected_license_header.split("\n")
+
+    for idx, line in enumerate(header_as_lst):
+        if re.search(copyright_pattern, line):
+            copyright_line += line
+            body = "\n".join(header_as_lst[idx + 1 :]).strip("#")
+            break
+        else:
+            copyright_line += line
+    copyright_line = copyright_line.replace("{YEAR}", r"20\d{2}").replace("{LICENSEE}", licensee)
+    appended_pattern = re.compile(rf"Copyright .* 20\d{{2}} {licensee}")
+
+    target = (
+        expected_license_header.replace("{YEAR}", r"20\d{2}")
+        .replace("{LICENSEE}", licensee)
+        .replace("\n", "")
+        .replace("(", r"\(")
+        .replace(")", r"\)")
+        .replace(".", r"\.")
+        .replace("'", r"\'")
+        .replace('"', r"\"")
+        .strip()
+    )
+
+    pattern = re.compile(target)
+    valid = False
+
+    for license_header in license_header_lst:
+        similar_body = (
+            difflib.SequenceMatcher(None, body, license_header.license_header).ratio() > 0.94
```

Comment:
we also might want to generalize this a bit to determine similarity, e.g. by comparing
```python
    "".join(line.lstrip("#").strip().lower() for line in header.splitlines()),
    "".join(line.lstrip("#").strip().lower() for line in expected_header.splitlines()),
```
so that licenses will always get marked as similar if they only differ by cases/whitespace/comment style/etc